tests/jest/cursors.js

```diff
@@ -0,0 +1,130 @@
+const helpers = require("./helpers/main")
+
+// Each test in this group will have
+// * 2 clients for alice (alice and aliceMobile)
+// * 1 room with Alice as the only member (room)
+
+// Cursors point to the last read message (by the user) in a room.
+// Cursors are tightly coupled with unread counts. The unread count
+// for a room is the difference the number of messages sent in the
+// room after the current read cursor.
+
+describe("A read cursor", () => {
+  test("is undefined if not set", async () => {
+    const actual = await page.evaluate(async () => {
+      const alice = await aliceChatManager.connect()
+      return alice.readCursor
+    })
+
+    expect(actual).toBeUndefined
+  })
+
+  test("invokes hook when set on other device", async () => {
+    // cursor positions should be message ids, but any int is valid
+    const expectedPos = 42
+
+    await page.evaluate(
+      async (roomId, expectedPos) => {
+        window.actual = undefined
+
+        // set up Alice's hook to set the global value
+        await aliceChatManager.connect({
+          onNewReadCursor: cursor => {
+            window.actual = cursor
+          },
+        })
+
+        const aliceMobile = await aliceMobileChatManager.connect()
+        await aliceMobile.setReadCursor({
+          roomId: roomId,
+          position: expectedPos,
+        })
+        return
+      },
+      room.id,
+      expectedPos,
+    )
+
+    // wait for hook to be invoked
+    await helpers.withHook(actual => {
+      expect(actual.position).toBe(expectedPos)
+    })
+  })
+
+  test("sets unread count", async () => {
+    // before a cursor is set, the unread count is undefined
+    const initial = await page.evaluate(async () => {
+      const alice = await aliceChatManager.connect()
+      return {
+        unread: alice.rooms[0].unreadCount,
+        messageAt: alice.rooms[0].lastMessageAt,
+      }
+    })
+    expect(initial.unread).toBeUndefined
+    expect(initial.messageAt).toBeUndefined
+
+    res = await helpers.makeSimpleMessage({
+      roomId: room.id,
+      userId: alice.id,
+      text: "hi",
+    })
+
+    // setting the cursor to the latest message sets the unread
+    // count to 0
+    await page.evaluate(
+      async (roomId, messageId) => {
+        window.actual = undefined
+        const alice = await aliceChatManager.connect({
+          onRoomUpdated: room =>
```

Comment:
There's a race condition here because if you connect fast enough you might be told about the unread count change from publishing a message as well as from setting a cursor. Probably want to set this hook first, then send a message, check you get the update, then set cursor, check you get the second update.